Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1557,17 +1557,49 @@ extension BrowserViewController: URLBarDelegate {
     }
 
     func urlBar(urlBar: URLBarView, didSubmitText text: String) {
-        // If we can't make a valid URL, do a search query.
-        // If we still don't have a valid URL, something is broken. Give up.
-        let engine = profile.searchEngines.defaultEngine
-        guard let url = URIFixup.getURL(text) ??
-                        engine.searchURLForQuery(text) else {
-            log.error("Error handling URL entry: \"\(text)\".")
+        let trimmedText = text.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceCharacterSet())
+        
+        guard trimmedText.contains(" ") else {
+            submitText(text)
             return
         }
 
-        Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        profile.bookmarks.modelFactory >>== {
+            let possibleKeywordQuerySeparatorSpace = trimmedText.characters.indexOf(" ")!
+            let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+            let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+            
+            $0.getURLForKeywordSearch(possibleKeyword)
+            .uponQueue(dispatch_get_main_queue()) {
+                if let urlString = $0.successValue {
```

Comment:
I've noticed that this line sometimes gets hit twice (even though I haven't re-edited and re-submitted the text). Haven't been able to reproduce that again.